openupgradelib/openupgrade_merge_records.py

```diff
@@ -682,6 +682,53 @@ def _delete_records_orm(env, model_name, record_ids, target_record_id):
         )
 
 
+def _check_no_recurrence(env, model_name, record_ids, target_record_id,
+                         model_table=None):
+    if not model_table:
+        model_table = env[model_name]._table
+    env.cr.execute("""
+        SELECT tc.table_name, kcu.column_name, COALESCE(imf.column1, 'id')
+        FROM information_schema.table_constraints AS tc
+        JOIN information_schema.key_column_usage AS kcu
+            ON tc.constraint_name = kcu.constraint_name
+            AND tc.table_schema = kcu.table_schema
+        JOIN information_schema.constraint_column_usage AS ccu
+            ON ccu.constraint_name = tc.constraint_name
+            AND ccu.table_schema = tc.table_schema
+        JOIN ir_model_fields AS imf
+            ON imf.model = %s AND imf.relation = imf.model AND ((
+                imf.name = kcu.column_name AND
+                tc.table_name = ccu.table_name) OR (
+                imf.column2 = kcu.column_name AND
+                tc.table_name = imf.relation_table))
+        WHERE tc.constraint_type = 'FOREIGN KEY'
+        AND ccu.table_name = %s and ccu.column_name = 'id'
+        """, (model_name, model_table))
+    for table, column, origin in env.cr.fetchall():
+        query = sql.SQL(
+            """SELECT {column} FROM {table}
+            WHERE {origin} = %(target_record_id)s"""
+        ).format(
+            table=sql.Identifier(table), column=sql.Identifier(column),
+            origin=sql.Identifier(origin),
+        )
+        env.cr.execute(query, {
+            'target_record_id': target_record_id,
+        })
+        new_parent_row = env.cr.fetchall()
+        if new_parent_row and new_parent_row[0] in record_ids:
+            # When we already have recursive hierarchy, doing a
+            # merge of a parent into one of their children let the
+            # awkward situation where the child points to itself,
+            # so we avoid it checking this condition
+            logger.info(
+                "Couldn't merge %s record(s) of model %s to record_id %s"
+                " to avoid recursion with field %s of table %s",
+                len(record_ids), model_name, target_record_id, origin, table)
+            return True
```

Comment:
Thanks for the refactoring. Now I'm just concerned with the semantics: I would expect a method called "check_no_recurrence" to return True if there is indeed "no recurrence" but I believe that is the other way around here, right?

Reply:
hahaha Ok, I renamed it to "_check_recurrence"